wcfsetup/install/files/lib/data/I18nDatabaseObjectList.class.php

```diff
@@ -0,0 +1,65 @@
+<?php
+namespace wcf\data;
+use wcf\system\exception\SystemException;
+use wcf\system\WCF;
+use wcf\util\StringUtil;
+
+/**
+ * Abstract class for a list of database objects with better sorting of i18n-based columns.
+ *
+ * @author	Florian Gail
+ * @copyright	2001-2017 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\Data
+ * @since	3.1
+ */
+abstract class I18nDatabaseObjectList extends DatabaseObjectList {
+	/**
+	 * Array of column names, that are eventually filled with language items.
+	 * Those additional joins are going to slow down the system; you should use this
+	 * class only when it's really needed.
+	 * The key represents the original field name.
+	 * The value represents the new field name containing the localized field-value.
+	 *
+	 * @example   [ 'title' => 'titleSortField' ]
+	 *
+	 * @var string[]
+	 */
+	public $i18nFields = [];
+	
+	/**
+	 * ID of the language that should be used
+	 * the value defaults to the current user's language
+	 *
+	 * @var integer
+	 */
+	public $languageID;
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function __construct() {
+		parent::__construct();
+		
+		if ($this->languageID === null) {
+			$this->languageID = WCF::getUser()->languageID;
+		}
+		
+		if (!empty($this->i18nFields)) {
+			foreach ($this->i18nFields as $key => $value) {
+				if (is_numeric($key)) {
+					throw new SystemException("'".$this->className."::\$i18nFields' should contain key-value pairs. '".$value."' has a no or a numeric key.");
```

Comment:
Please do not use `SystemException` but one of PHP's own exception classes, probably `\LogicException`. Also, `$this->className` contains the DBO class name, not the list name.